Return message string from model panic handler to be displayed in UI …

mxdamien opened the pull request:
…status

closes #1519

hannobraun commented on crates/fj/src/abi/context.rs:
```diff
@@ -23,7 +23,9 @@ impl<'a> From<&'a &dyn crate::models::Context> for Context<'a> {
                 Ok(None) => StringSlice::from_str(""),
-                Err(payload) => crate::abi::on_panic(payload),
+                Err(payload) => StringSlice::from_str(
+                    crate::abi::on_panic(payload).to_string().as_str(),
```
This is unsound: `on_panic` returns an `ffi_safe::String`, which is converted into a regular `String`, which is then dropped. The `StringSlice` that references it is returned from the function, meaning the the `StringSlice` outlives the `String` it references. Also see [`StringSlice`'s documentation](https://github.com/hannobraun/Fornjot/blob/71d16ce6fb3a6e612c0e89bf46eee7278a5d1c54/crates/fj/src/abi/ffi_safe.rs#L259-L262).

---

In addition, this returns the error message just like it would return the value of the parameter. If I read the code right, this error message is then parsed as a value in the expanded model code ([[1]](https://github.com/hannobraun/Fornjot/blob/71d16ce6fb3a6e612c0e89bf46eee7278a5d1c54/crates/fj-proc/src/expand.rs#L137), [[2]](https://github.com/hannobraun/Fornjot/blob/71d16ce6fb3a6e612c0e89bf46eee7278a5d1c54/crates/fj-proc/src/expand.rs#L145)).

I think this is semi-okay in most cases, as it will cause a parse error. It might still be confusing to the user though. But it gets worse, if the parameter is a `String`. Imagine you're making a part that you want to engrave a custom message into, and you are engraving the error message instead. We're entering JavaScript territory here :smile: 

---

I think we can solve both issues by doing the following:

1. Return a `Result` (probably [`ffi_safe::Result`](https://github.com/hannobraun/Fornjot/blob/71d16ce6fb3a6e612c0e89bf46eee7278a5d1c54/crates/fj/src/abi/ffi_safe.rs#L215-L221) is required) from this method. The `Err` variant can just be the `ffi_safe::String` returned from `on_panic`, for a start.
2. Adapt all methods that call this, as necessary, returning `Result` where needed. Specifically [`Context::get_argument`](https://github.com/hannobraun/Fornjot/blob/71d16ce6fb3a6e612c0e89bf46eee7278a5d1c54/crates/fj/src/abi/context.rs#L39-L56) probably needs a custom error type, to distinguish between "argument not found" and "panicked" cases.
3. The generated model code already handles parse errors (here are the links to that expand code again: [[1]](https://github.com/hannobraun/Fornjot/blob/71d16ce6fb3a6e612c0e89bf46eee7278a5d1c54/crates/fj-proc/src/expand.rs#L137), [[2]](https://github.com/hannobraun/Fornjot/blob/71d16ce6fb3a6e612c0e89bf46eee7278a5d1c54/crates/fj-proc/src/expand.rs#L145)). We can adapt that code to handle errors resulting from a panic. ( I found it helpful to expand the macro code, for example with `cargo expand -p spacer` for the spacer model, to understand what's happening.)

hannobraun commented on crates/fj/src/abi/model.rs:
```diff
@@ -46,7 +46,11 @@ impl From<Box<dyn crate::models::Model>> for Model {
                 Ok(meta) => meta.into(),
-                Err(payload) => crate::abi::on_panic(payload),
+                Err(payload) => {
+                    model.metadata().name =
+                        crate::abi::on_panic(payload).into();
```
I think we should adapt this method to return a `Result` and return an `Err` if a panic happens. Just parking the error message in the metadata, where it might not be noticed, is way too error-prone.

hannobraun reviewed (changes requested):
Thank for you for this pull request, @mxdamien!

I think this is a good start. The main cause of panics, the user-written model code, is taken care of. However, there are some problems around the periphery. I've left some comments with my thoughts. Please let me know what you think!

Also, and this is more a note to myself, here's what a `todo!()` in model code looks like, as of this pull request:
![Screenshot from 2023-01-24 13-07-23](https://user-images.githubusercontent.com/85732/214287863-790012bd-bf3d-4e10-8898-ce286cd5cbb4.png)

I think this is good enough for now, but eventually we would want an error message that makes clear that the error originated from a panic in the model code, and add as much detail as possible (file name, line, etc.).

None of this needs to be a part of this pull request. I'm just making a note, so I remember to open an issue, once this is merged.

hannobraun commented:
Thank you, David!

mxdamien commented:
Hi !

I hope I've addressed your remarks. I'm not entirely sure about the usage of the different internal / ffi-boundary structs. It's my first time dealing with this kind of stuff in Rust.

David

hannobraun commented:
Thank you for keeping at it, David! I'll take a look later today.

hannobraun commented on crates/fj/src/abi/context.rs:
```diff
@@ -47,9 +61,17 @@ impl crate::models::Context for Context<'_> {
+                            false => Some(other.into_str()),
+                        }
+                    }
+                    super::ffi_safe::Result::Err(_) => None,
```
It's not ideal to swallow the error here, as that could cause bugs to be hidden. Instead of getting an error message, making it clear that a bug has occurred, it could lead to a specified argument just not being recognized, which would be very confusing.

I think `Context::get_argument` should also return `Result` instead of an `Option`.

hannobraun reviewed (approved):
Thank you, @mxdamien! This is looking great!

I've left one comment about a thing I don't like, but I decided to merge the pull request anyway. Panics in the model code are properly handled now, and I believe that the error being swallowed there is a corner case.

I'm going to open an issue to track that. Feel free to follow up with another pull request to address it, if you want.
